File: src/adventofcode2018/Day15.java
package adventofcode2018;

import java.awt.Point;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.function.Predicate;
import java.util.stream.Collectors;

import utils.Input;

/**
 * https://adventofcode.com/2018/day/15
 */
public class Day15 {

    static class Unit {
        int x, y;
        int hp = 200;
        int type;
        Unit(int x, int y, int type) {
            this.x = x;
            this.y = y;
            this.type = type;
        }
    }

    static int opposite(int type) {
        return type == 'E' ? 'G' : 'E';
    }

    static final int[][] dirs = {{0, -1}, {-1, 0}, {1, 0}, {0, 1}};

    static Point wave(int[][] map, Predicate<Point> pred, Point start) {
        List<Point> points = Collections.singletonList(start);
        Set<Point> visited = new HashSet<>();
        Point target = null;
        // find nearest point that satisfies the condition
        while (!points.isEmpty() && target == null) {
            List<Point> nextPoints = new ArrayList<>();
            for (Point p : points) {
                for (int dir[] : dirs) {
                    Point p2 = new Point(p.x+dir[0], p.y+dir[1]);
                    if (visited.contains(p2)) {
                        continue;
                    }
                    if (pred.test(p2)) {
                        if (target == null || p.y < target.y || (p.y == target.y && p.x < target.x)) {
                            target = p;
                        }
                    }
                    if (map[p2.y][p2.x] == '.') {
                        nextPoints.add(p2);
                        visited.add(p2);
                    }
                }
            }
            points = nextPoints;
        }
        return target;
    }

    public static void main(String[] args) {
        long time = System.currentTimeMillis();
        List<String> results = new ArrayList<>();
        for (int power = 3; ; power++) {
            int t = 0;
            int powerFinal = power;
            int[][] map = new Input(2018, 15).getMap();
            List<Unit> units = new ArrayList<>();
            // collect all the units
            for (int y=0; y<map.length; y++) {
                for (int x=0; x<map[y].length; x++) {
                    if (map[y][x] == 'E' || map[y][x] == 'G') {
                        units.add(new Unit(x, y, map[y][x]));
                    }
                }
            }
            for (;/* rounds */;) {
                boolean end = true;
                units.sort(Comparator.comparing(u -> u.y*1000 + u.x));
                for (Unit u : units) {
                    if (u.hp <= 0) continue;
                    // find nearest target
                    Point target = wave(map,
                            p -> map[p.y][p.x] == opposite(u.type),
                            new Point(u.x, u.y));
                    if (target == null) {
                        continue;
                    }
                    end = false;
                    // move if needed
                    if (target.x != u.x || target.y != u.y) {
                        target = wave(map, p -> p.x == u.x && p.y == u.y, target);
                        map[u.y][u.x] = '.';
                        map[target.y][target.x] = u.type;
                        u.x = target.x;
                        u.y = target.y;
                    }
                    // atack if possible
                    Arrays.stream(dirs)
                            .filter(dir -> map[u.y+dir[1]][u.x+dir[0]] == opposite(u.type))
                            .map(dir -> units.stream()
                                    .filter(u2 -> u2.x == u.x+dir[0] && u2.y == u.y+dir[1] && u2.hp > 0)
                                    .findFirst().get())
                            .sorted(Comparator.comparing(u2 -> u2.hp * 10000 + u2.y * 100 + u2.x))
                            .findFirst().ifPresent(u2 -> {
                                u2.hp -= u.type == 'E' ? powerFinal : 3;
                                if (u2.hp <= 0) map[u2.y][u2.x] = '.';
                            });
                }
                if (end) {
                    break;
                }
                t++;
            }
            int sumHP = units.stream().mapToInt(u -> u.hp).filter(u -> u > 0).sum();
            results.add(String.format("power = %2d: %d * %d = %d", power, t-1, sumHP, (t-1) * sumHP));
            if (units.stream().noneMatch(u -> u.type == 'E' && u.hp <= 0)) {
                break;
            }
        }
        System.err.println(results.get(0));
        System.err.println(results.get(results.size() - 1));
        System.out.println("time: " + (System.currentTimeMillis() - time));
    }

    static void printMap(int[][] map) {
        Arrays.stream(map)
                .map(s -> Arrays.stream(s).mapToObj(c -> Character.toString((char)c))
                        .collect(Collectors.joining()))
                .forEach(System.out::println);
        System.out.println();
    }
}

File: src/Python/2018/Day15.py

# https://adventofcode.com/2018/day/15

class Unit:
    def __init__(self, x, y, type):
        self.x, self.y = x, y
        self.hp = 200
        self.type = type

    def opposite(self):
        return "E" if self.type == "G" else "G"

dirs = [[0, -1], [-1, 0], [1, 0], [0, 1]]

def wave(field, predicate, start):
    points = [start]
    visited = set()
    while len(points) > 0:
        nextPoints = []
        for p in sorted(points, key = lambda p: p[1] * 100 + p[0]):
            for dir in dirs:
                p2 = (p[0] + dir[0], p[1] + dir[1])
                if p2 in visited:
                    continue
                if predicate(p2):
                    return p
                if field[p2[1]][p2[0]] == ".":
                    nextPoints.append(p2)
                    visited.add(p2)
        points = nextPoints

def simulate(field, power):
    t = 0
    units = []
    # collect units
    for y in range(len(field)):
        for x in range(len(field[y])):
            if field[y][x] == "E" or field[y][x] == "G":
                units.append(Unit(x, y, field[y][x]))
    # simulate the buttle
    end = False
    while not end:
        end = True
        units.sort(key = lambda u: u.y*1000 + u.x)
        for u in units:
            if u.hp <= 0: continue
            # finding a target
            target = wave(field, lambda p: field[p[1]][p[0]] == u.opposite(), (u.x, u.y))
            if not target: continue
            end = False
            # move
            if target != (u.x, u.y):
                target = wave(field, lambda p: (p[0], p[1]) == (u.x, u.y), target)
                field[u.y][u.x] = '.'
                field[target[1]][target[0]] = u.type
                u.x, u.y = target
            # attack
            targets = []
            for dir in dirs:
                if field[u.y + dir[1]][u.x + dir[0]] == u.opposite():
                    targets.append(next(filter(lambda u2: (u2.x, u2.y) == (u.x + dir[0], u.y + dir[1]) and u2.hp > 0, units)))
            if len(targets) > 0:
                u2 = min(targets, key = lambda u: u.hp * 10000 + u.y * 100 + u.x)
                u2.hp -= power if u.type == "E" else 3
                if u2.hp <= 0:
                    field[u2.y][u2.x] = "."
        t += 1
    return (t - 1, units)

from utils import Input, printResult
import copy

results = []
field = list(map(list, Input(2018, 15).lines()))
for power in range(3, 100):
    t, units = simulate(copy.deepcopy(field), power)
    sumHP = sum(u.hp for u in units if u.hp > 0)
    results.append("power = %d, %d * %d = %d" % (power, t-1, sumHP, (t-1) * sumHP))
    # All elves are alife
    if not any(u.type == "E" and u.hp <= 0 for u in units):
        break

printResult(1, results[0])
printResult(2, results[-1])
